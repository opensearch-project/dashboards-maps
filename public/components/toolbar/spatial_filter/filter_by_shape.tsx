/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiButtonIcon, EuiContextMenu, EuiPanel, EuiPopover } from '@elastic/eui';
import React, { useState } from 'react';
import {
  DRAW_FILTER_CANCEL,
  DRAW_FILTER_SHAPE_TITLE,
  DRAW_FILTER_SPATIAL_RELATIONS,
  DrawFilterProperties,
  FILTER_DRAW_MODE,
} from '../../../../common';
import { FilterInputPanel } from './filter_input_panel';

interface FilterByShapeProps {
  setDrawFilterProperties: (properties: DrawFilterProperties) => void;
  mode: FILTER_DRAW_MODE;
  shapeMode: FILTER_DRAW_MODE;
  shapeLabel: string;
  defaultLabel: string;
  iconType: any;
}

export const FilterByShape = ({
  shapeMode,
  setDrawFilterProperties,
  mode,
  defaultLabel,
  iconType,
  shapeLabel,
}: FilterByShapeProps) => {
  const [isPopoverOpen, setPopover] = useState(false);
  const isFilterActive: boolean = mode === shapeMode;

  const onClick = () => {
    if (isFilterActive) {
      setDrawFilterProperties({
        mode: FILTER_DRAW_MODE.NONE,
      });
      return;
    }
    setPopover(!isPopoverOpen);
  };

  const closePopover = () => {
    setPopover(false);
  };

  const onSubmit = (input: { relation: string; label: string; mode: FILTER_DRAW_MODE }) => {
    setDrawFilterProperties({
      mode: input.mode,
      relation: input.relation,
      filterLabel: input.label,
    });
    closePopover();
  };

  const panels = [
    {
      id: 0,
      title: DRAW_FILTER_SHAPE_TITLE,
      content: (
        <FilterInputPanel
          drawLabel={shapeLabel}
          defaultFilterLabel={defaultLabel}
          relations={DRAW_FILTER_SPATIAL_RELATIONS}
          onSubmit={onSubmit}
          mode={shapeMode}
        />
      ),
    },
  ];

  const drawShapeButton = (
    <EuiPanel paddingSize="none" className="spatialFilterToolbar__shape">
      <EuiButtonIcon
        isSelected={isFilterActive}
        display={isFilterActive ? 'fill' : 'empty'}
        aria-pressed={isFilterActive}
        color="primary"
        size={'s'}
        iconType={iconType}
        onClick={onClick}
        aria-label={'draw_filter_shape'}
        title={isFilterActive ? DRAW_FILTER_CANCEL : shapeLabel}
      />
    </EuiPanel>
  );
  return (
    <EuiPopover
      id="drawShapeId"
      button={drawShapeButton}
      isOpen={isPopoverOpen}
      closePopover={closePopover}
      panelPaddingSize="none"
      anchorPosition="leftUp"
      data-test-subj="drawShapePopOver"
    >
      <EuiContextMenu initialPanelId={0} panels={panels} />
    </EuiPopover>
  );
};
