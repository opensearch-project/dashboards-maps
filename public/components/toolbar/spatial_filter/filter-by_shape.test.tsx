/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import renderer from 'react-test-renderer';
import React from 'react';
import { FilterByShape } from './filter_by_shape';
import {
  DRAW_FILTER_POLYGON,
  DRAW_FILTER_POLYGON_DEFAULT_LABEL,
  DRAW_FILTER_RECTANGLE,
  DRAW_FILTER_RECTANGLE_DEFAULT_LABEL,
  FILTER_DRAW_MODE,
} from '../../../../common';

describe('render polygon', function () {
  it('renders filter by polygon option', () => {
    const mockCallback = jest.fn();
    const polygonComponent = renderer
      .create(
        <FilterByShape
          setDrawFilterProperties={mockCallback}
          mode={FILTER_DRAW_MODE.NONE}
          shapeMode={FILTER_DRAW_MODE.POLYGON}
          shapeLabel={DRAW_FILTER_POLYGON}
          defaultLabel={DRAW_FILTER_POLYGON_DEFAULT_LABEL}
          iconType={'vector'}
        />
      )
      .toJSON();
    expect(polygonComponent).toMatchSnapshot();
  });

  it('renders filter by polygon in middle of drawing', () => {
    const mockCallback = jest.fn();
    const polygonComponent = renderer
      .create(
        <FilterByShape
          setDrawFilterProperties={mockCallback}
          mode={FILTER_DRAW_MODE.POLYGON}
          shapeMode={FILTER_DRAW_MODE.POLYGON}
          shapeLabel={DRAW_FILTER_POLYGON}
          defaultLabel={DRAW_FILTER_POLYGON_DEFAULT_LABEL}
          iconType={'vector'}
        />
      )
      .toJSON();
    expect(polygonComponent).toMatchSnapshot();
  });
});

describe('render rectangle', function () {
  it('renders filter by rectangle option', () => {
    const mockCallback = jest.fn();
    const rectangle = renderer
      .create(
        <FilterByShape
          setDrawFilterProperties={mockCallback}
          mode={FILTER_DRAW_MODE.NONE}
          shapeMode={FILTER_DRAW_MODE.RECTANGLE}
          shapeLabel={DRAW_FILTER_RECTANGLE}
          defaultLabel={DRAW_FILTER_RECTANGLE_DEFAULT_LABEL}
          iconType={'vector'}
        />
      )
      .toJSON();
    expect(rectangle).toMatchSnapshot();
  });

  it('renders filter by rectangle in middle of drawing', () => {
    const mockCallback = jest.fn();
    const rectangle = renderer
      .create(
        <FilterByShape
          setDrawFilterProperties={mockCallback}
          mode={FILTER_DRAW_MODE.RECTANGLE}
          shapeMode={FILTER_DRAW_MODE.RECTANGLE}
          shapeLabel={DRAW_FILTER_RECTANGLE}
          defaultLabel={DRAW_FILTER_RECTANGLE_DEFAULT_LABEL}
          iconType={'vector'}
        />
      )
      .toJSON();
    expect(rectangle).toMatchSnapshot();
  });
});
