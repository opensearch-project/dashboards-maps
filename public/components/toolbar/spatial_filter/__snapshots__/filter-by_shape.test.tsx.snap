// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render polygon renders filter by polygon in middle of drawing 1`] = `
<div
  className="euiPopover euiPopover--anchorLeftUp"
  data-test-subj="drawShapePopOver"
  id="drawShapeId"
>
  <div
    className="euiPopover__anchor"
  >
    <div
      className="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow spatialFilterToolbar__shape"
    >
      <button
        aria-label="draw_filter_shape"
        aria-pressed={true}
        className="euiButtonIcon euiButtonIcon--primary euiButtonIcon--fill euiButtonIcon--small"
        disabled={false}
        onClick={[Function]}
        title="Cancel"
        type="button"
      >
        <svg
          aria-hidden={true}
          className="euiIcon euiIcon--medium euiIcon--inherit euiButtonIcon__icon"
          focusable="false"
          height={16}
          role="img"
          style={null}
          viewBox="0 0 16 16"
          width={16}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.5 11V5H11V3.5H5V5H3.5v6H5v1.5h6V11h1.5zm1 0H15v4h-4v-1.5H5V15H1v-4h1.5V5H1V1h4v1.5h6V1h4v4h-1.5v6zM4 4V2H2v2h2zm8 0h2V2h-2v2zM2 14h2v-2H2v2zm10 0h2v-2h-2v2z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`render polygon renders filter by polygon option 1`] = `
<div
  className="euiPopover euiPopover--anchorLeftUp"
  data-test-subj="drawShapePopOver"
  id="drawShapeId"
>
  <div
    className="euiPopover__anchor"
  >
    <div
      className="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow spatialFilterToolbar__shape"
    >
      <button
        aria-label="draw_filter_shape"
        aria-pressed={false}
        className="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--small"
        disabled={false}
        onClick={[Function]}
        title="Draw Polygon"
        type="button"
      >
        <svg
          aria-hidden={true}
          className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
          focusable="false"
          height={16}
          role="img"
          style={null}
          viewBox="0 0 16 16"
          width={16}
          xmlns="http://www.w3.org/2000/svg"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`render rectangle renders filter by rectangle in middle of drawing 1`] = `
<div
  className="euiPopover euiPopover--anchorLeftUp"
  data-test-subj="drawShapePopOver"
  id="drawShapeId"
>
  <div
    className="euiPopover__anchor"
  >
    <div
      className="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow spatialFilterToolbar__shape"
    >
      <button
        aria-label="draw_filter_shape"
        aria-pressed={true}
        className="euiButtonIcon euiButtonIcon--primary euiButtonIcon--fill euiButtonIcon--small"
        disabled={false}
        onClick={[Function]}
        title="Cancel"
        type="button"
      >
        <svg
          aria-hidden={true}
          className="euiIcon euiIcon--medium euiIcon--inherit euiButtonIcon__icon"
          focusable="false"
          height={16}
          role="img"
          style={null}
          viewBox="0 0 16 16"
          width={16}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.5 11V5H11V3.5H5V5H3.5v6H5v1.5h6V11h1.5zm1 0H15v4h-4v-1.5H5V15H1v-4h1.5V5H1V1h4v1.5h6V1h4v4h-1.5v6zM4 4V2H2v2h2zm8 0h2V2h-2v2zM2 14h2v-2H2v2zm10 0h2v-2h-2v2z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`render rectangle renders filter by rectangle option 1`] = `
<div
  className="euiPopover euiPopover--anchorLeftUp"
  data-test-subj="drawShapePopOver"
  id="drawShapeId"
>
  <div
    className="euiPopover__anchor"
  >
    <div
      className="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow spatialFilterToolbar__shape"
    >
      <button
        aria-label="draw_filter_shape"
        aria-pressed={false}
        className="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--small"
        disabled={false}
        onClick={[Function]}
        title="Draw Rectangle"
        type="button"
      >
        <svg
          aria-hidden={true}
          className="euiIcon euiIcon--medium euiIcon--inherit euiButtonIcon__icon"
          focusable="false"
          height={16}
          role="img"
          style={null}
          viewBox="0 0 16 16"
          width={16}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.5 11V5H11V3.5H5V5H3.5v6H5v1.5h6V11h1.5zm1 0H15v4h-4v-1.5H5V15H1v-4h1.5V5H1V1h4v1.5h6V1h4v4h-1.5v6zM4 4V2H2v2h2zm8 0h2V2h-2v2zM2 14h2v-2H2v2zm10 0h2v-2h-2v2z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
